bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -437,12 +437,21 @@ func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Contex
 	}
 
 	certificates := secret.NewCertificatesForInitialControlPlane(scope.Config.Spec.ClusterConfiguration)
-	err = certificates.LookupOrGenerate(
-		ctx,
+
+	// Find the certificates associated with the Cluster.
+	err = certificates.Lookup(ctx,
 		r.Client,
-		util.ObjectKey(scope.Cluster),
-		*metav1.NewControllerRef(scope.Config, bootstrapv1.GroupVersion.WithKind("KubeadmConfig")),
-	)
+		util.ObjectKey(scope.Cluster))
+
+	// If the Cluster does not have a ControlPlaneReference look up and generate the certificates.
+	// If there is a ControlPlane ref look up the existing certificates.
+	if scope.Cluster.Spec.ControlPlaneRef == nil {
```

Comment:
Could there be an impactful race condition here with ClusterClass managed clusters? The ref won't be set on creation (and may take some time to be created if the templates are invalid etc.) but I don't think any Cluster-owned KubeadmConfig can be reconciled before this point.

Reply:
Could there be an impactful race condition here with ClusterClass managed clusters?

No, because machines (and related KubeadmConfig) are not created before the Cluster has a controlPlane and infrastructure ref